### DISCUSSION TITLE
Is it best to extract features before calling trim_to_supervision?

### DISCUSSION THREAD

**Discussion by RuABraun** [lhotse-speech/lhotse#1191] [Q&A] [answered]
It seems to me after trim_to_supervision() doing feature extraction will be quite inefficient because to extract the features the audio files will be loaded repeatedly for every cut (instead of say loading once and then extracting features for all the supervisions).

So my question is is my thinking correct that it's best to extract features and then call trim_to_supervision?

And I guess trim_to_supervision is not reversible in some way (going back to all supervisions of one audio file being in one cut)?

**Comment by desh2608** [comment, marked as answer]
Lhotse only loads the relevant segments of the audio, so the full recording is not loaded. Usually I have 2 rules of thumb when deciding whether to trim before or after feature extraction:

* If I will be using different segmentations of the same audio, I extract features *before*. This is so that all the segmentations can use the same underlying features. For example, I used this approach recently when working on long-form decoding with different segment sizes.
* If the supervisions are only referring to a small part of the full recording, I extract *after* trimming. This is the case for some corpora such as VoxPopuli, where the full recordings are several hundred hours, whereas the transcriptions are much shorter.

**Reply by RuABraun** [reply]
I should have mentioned the files are urls.

I knew about the partial loading but doing a random seek can't be ideal (as I understand that is the expensive part, not loading the actual data on disk), therefore me wondering whether trim_to_supervisions is meant to be called after feat extraction.

Thank you for sharing your approach that's useful to know!

**Reply by desh2608** [reply]
Ahh, yes. If the audio source is a URL or a bash pipe command, it is better to just load it once to compute features.
